File: src/HeadersInfo.java
import java.security.SecureRandom;
import java.util.HashMap;
import java.util.Map;
import java.util.Random;
import java.util.UUID;

public class HeadersInfo {
    private static final String HEX_CHARS = "0123456789abcdef";
    private static final SecureRandom secureRandom = new SecureRandom();
    public static final String VScode_SessionId = UUID.randomUUID() + String.valueOf((long)(Math.random() * 1e13));
    public static final String vscode_machineid = generateRandomHex(64);
    public static final String openai_organization = "github-copilot";
    public static final String editor_version = "vscode/1.98.0-insider";
    public static final String copilot_language_server_version = "0.23.2";
    public static final String editor_plugin_version = "copilot-chat/"+copilot_language_server_version;




    public static final String x_github_api_version = "2025-01-21";
    public static final String content_type = "application/json";
    public static final String user_agent = "GitHubCopilotChat/"+copilot_language_server_version;

    /**
     * Build request headers
     */
    public static Map<String, String> getCopilotHeaders(String longTermToken) {
        // 先从 DB 读取 machine_id
        TokenManager tokenManager = new TokenManager();
        String machineId = null;

        try {
            machineId = tokenManager.getMachineId(longTermToken);
            if (machineId == null) {
                // 不存在则生成并尝试写入
                machineId = generateRandomHex(64);
                boolean saved = tokenManager.setMachineId(longTermToken, machineId);
                if (!saved) {
                    System.out.println("Warning: failed to save new machine_id, using generated one.");
                }
            }
        } catch (Exception e) {
            // 读取或写入失败，退回到新生成的
            System.out.println("Error accessing machine_id in DB: " + e.getMessage());
            machineId = generateRandomHex(64);
        }
        // 生成一个 sessionId（每次启动不同）
        Map<String, String> headers = new HashMap<>();
        headers.put("Content-Type", content_type);
        headers.put("Connection", "keep-alive");
        headers.put("Editor-Plugin-Version", editor_plugin_version);
        headers.put("copilot-integration-id","vscode-chat");
        headers.put("Editor-Version", editor_version);
        headers.put("Openai-Organization", openai_organization);
        headers.put("User-Agent", user_agent);
        headers.put("VScode-MachineId", machineId);
        headers.put("VScode-SessionId", VScode_SessionId);
        headers.put("accept", "*/*");
        headers.put("Sec-Fetch-Site", "none");
        headers.put("Sec-Fetch-Mode", "no-cors");
        headers.put("Sec-Fetch-Dest", "empty");
        headers.put("accept-encoding", "gzip, deflate, br, zstd");
        headers.put("X-GitHub-Api-Version", x_github_api_version);
        headers.put("X-Request-Id", RandomXRequestID(32));
        return headers;
    }

    public static String generateRandomHex(int length) {
        TokenManager tokenManager = new TokenManager();
        StringBuilder sb = new StringBuilder(length);
        for (int i = 0; i < length; i++) {
            int index = secureRandom.nextInt(HEX_CHARS.length());
            sb.append(HEX_CHARS.charAt(index));
        }
        return sb.toString();
    }

    public static String RandomXRequestID(int length) {
        StringBuilder sb = new StringBuilder(32);
        Random random = new Random();

        // Generate a 32-character random string consisting of digits and lowercase letters
        for (int i = 0; i < length; i++) {
            int index = random.nextInt("abcdefghijklmnopqrstuvwxyz0123456789".length());
            sb.append("abcdefghijklmnopqrstuvwxyz0123456789".charAt(index));
        }

        // Format the output as xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxxx
        String x = String.format("%s-%s-%s-%s-%s",
                sb.substring(0, 8),
                sb.substring(8, 12),
                sb.substring(12, 16),
                sb.substring(16, 20),
                sb.substring(20, 32));
        return x;
    }
}

File: go/headers_info.go
package main

import (
	"fmt"
	"math/rand"
	"net/http"
	"time"
)

const (
	openai_organization             = "github-copilot"
	editor_version                  = "vscode/1.98.0-insider"
	editor_plugin_version           = "copilot/1.270.0"
	copilot_language_server_version = "1.270.0"
	x_github_api_version            = "2025-01-21"
	user_agent                      = "GitHubCopilotChat/0.23.2"
)

// setCORSHeaders 设置 CORS 响应头
func setCORSHeaders(w http.ResponseWriter) {
	w.Header().Set("Access-Control-Allow-Origin", "*")
	w.Header().Set("Access-Control-Allow-Credentials", "true")
	w.Header().Set("Access-Control-Allow-Methods", "GET, POST, OPTIONS")
	w.Header().Set("Access-Control-Allow-Headers", "Content-Type, Authorization")
}

// GetCopilotHeaders 构造发送给 Copilot API 的请求头
func GetCopilotHeaders() map[string]string {
	headers := make(map[string]string)
	headers["Content-Type"] = "application/json"
	headers["Connection"] = "keep-alive"
	headers["Editor-Plugin-Version"] = editor_plugin_version
	headers["Editor-Version"] = editor_version
	headers["Openai-Organization"] = openai_organization
	headers["User-Agent"] = user_agent
	headers["VScode-MachineId"] = generateRandomHex(64)
	headers["VScode-SessionId"] = uuid()
	headers["accept"] = "*/*"
	headers["Sec-Fetch-Site"] = "none"
	headers["Sec-Fetch-Mode"] = "no-cors"
	headers["Sec-Fetch-Dest"] = "empty"
	headers["accept-encoding"] = "gzip, deflate, br, zstd"
	headers["X-GitHub-Api-Version"] = x_github_api_version
	headers["X-Request-Id"] = randomXRequestID(32)
	return headers
}

func generateRandomHex(n int) string {
	const hexChars = "0123456789abcdef"
	s := ""
	for i := 0; i < n; i++ {
		s += string(hexChars[rand.Intn(len(hexChars))])
	}
	return s
}

func uuid() string {
	return fmt.Sprintf("%x", rand.Int63())
}

func randomXRequestID(length int) string {
	const chars = "abcdefghijklmnopqrstuvwxyz0123456789"
	s := ""
	for i := 0; i < length; i++ {
		s += string(chars[rand.Intn(len(chars))])
	}
	if len(s) >= 32 {
		return fmt.Sprintf("%s-%s-%s-%s-%s", s[0:8], s[8:12], s[12:16], s[16:20], s[20:32])
	}
	return s
}

func init() {
	// 确保随机数种子初始化
	rand.Seed(time.Now().UnixNano())
}
